File: src/magnet_pattern_match.py
"""Match magnet names / file names using {CODE} / {code} placeholder rules."""

from __future__ import annotations

import re

CODE_UPPER = "{CODE}"
CODE_LOWER = "{code}"

_TOKEN_SPLIT = re.compile(r"(\{CODE\}|\{code\})")


def split_normalized_code(code: str) -> tuple[str, str]:
    """Return (upper_prefix_code, lower_prefix_code) e.g. IPZZ-576, ipzz-576."""
    normalized = str(code or "").upper().strip()
    match = re.match(r"^([A-Z]+)-?(\d+[A-Z]?)$", normalized)
    if not match:
        lowered = normalized.lower()
        return normalized, lowered
    upper = f"{match.group(1).upper()}-{match.group(2)}"
    lower = f"{match.group(1).lower()}-{match.group(2)}"
    return upper, lower


def normalize_code(code: str) -> str:
    return split_normalized_code(code)[0]


def _literal_to_regex(fragment: str) -> str:
    parts: list[str] = []
    for char in fragment:
        if char.isalpha():
            parts.append(f"[{char.lower()}{char.upper()}]")
        else:
            parts.append(re.escape(char))
    return "".join(parts)


def _code_token_to_regex(code: str) -> str:
    upper, _ = split_normalized_code(code)
    return _literal_to_regex(upper)


def pattern_to_regex(pattern: str, code: str) -> re.Pattern[str] | None:
    """Build anchored regex: {CODE}/{code} are equivalent; code letters are case-insensitive."""
    text = str(pattern or "").strip()
    if not text:
        return None

    body_parts: list[str] = []
    for token in _TOKEN_SPLIT.split(text):
        if token in (CODE_UPPER, CODE_LOWER):
            body_parts.append(_code_token_to_regex(code))
        elif token:
            body_parts.append(_literal_to_regex(token))

    body = re.sub(r"\s+", "", "".join(body_parts))
    if not body:
        return None
    return re.compile(f"^{body}$")


def matches_pattern(text: str, pattern: str, code: str) -> bool:
    """
    Full-string match. {CODE}-C matches IPZZ-576-c but not IPZZ-576.
    {CODE} and {code} are equivalent; code letters match case-insensitively.
    """
    regex = pattern_to_regex(pattern, code)
    if regex is None:
        return False
    normalized = re.sub(r"\s+", "", str(text or ""))
    return bool(regex.match(normalized))


def expand_pattern_display(pattern: str, code: str) -> str:
    """Best-effort preview string for UI (uses upper/lower placeholders only)."""
    upper, lower = split_normalized_code(code)
    return (
        str(pattern or "")
        .replace(CODE_UPPER, upper)
        .replace(CODE_LOWER, lower)
    )

File: extension/content/magnet-pattern-match.js
/**
 * {CODE} / {code} 等价，番号字母大小写不敏感（SSIS-204 = ssis-204）。
 * 字面量中的英文字母（如 -C）大小写不敏感：-C = -c。
 * 须完整匹配：{CODE}-C 不匹配 {CODE}。
 */
(function () {
  "use strict";

  const TOKEN_SPLIT = /(\{CODE\}|\{code\})/g;

  function splitNormalizedCode(code) {
    const text = String(code || "").toUpperCase().trim();
    const match = text.match(/^([A-Z]+)-?(\d+[A-Z]?)$/i);
    if (!match) {
      const lowered = text.toLowerCase();
      return { upper: text, lower: lowered };
    }
    return {
      upper: `${match[1].toUpperCase()}-${match[2]}`,
      lower: `${match[1].toLowerCase()}-${match[2]}`,
    };
  }

  function normalizeCode(code) {
    return splitNormalizedCode(code).upper;
  }

  function literalToRegex(fragment) {
    let out = "";
    for (const ch of fragment) {
      if (/[a-zA-Z]/.test(ch)) {
        out += `[${ch.toLowerCase()}${ch.toUpperCase()}]`;
      } else {
        out += ch.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      }
    }
    return out;
  }

  function codeTokenToRegex(code) {
    const { upper } = splitNormalizedCode(code);
    return literalToRegex(upper);
  }

  function patternToRegex(pattern, code) {
    const text = String(pattern || "").trim();
    if (!text) return null;

    const parts = text.split(TOKEN_SPLIT).map((token) => {
      if (token === "{CODE}" || token === "{code}") return codeTokenToRegex(code);
      if (!token) return "";
      return literalToRegex(token);
    });

    const body = parts.join("").replace(/\s+/g, "");
    if (!body) return null;
    return new RegExp(`^${body}$`);
  }

  function matchesPattern(text, pattern, code) {
    const regex = patternToRegex(pattern, code);
    if (!regex) return false;
    const normalized = String(text || "").replace(/\s+/g, "");
    return regex.test(normalized);
  }

  const FOUR_K_PATTERNS = [
    /\[4k\]/i,
    /(?:^|[^a-z0-9])4k(?:[^a-z0-9]|$)/i,
    /\b2160p\b/i,
    /\buhd\b/i,
  ];

  function is4kMagnetName(display) {
    const text = String(display || "").trim();
    if (!text) return false;
    return FOUR_K_PATTERNS.some((pattern) => pattern.test(text));
  }

  window.JM_magnetPatternMatch = {
    splitNormalizedCode,
    normalizeCode,
    matchesPattern,
    patternToRegex,
    is4kMagnetName,
    FOUR_K_PATTERNS,
  };
})();
